crates/evm/src/query.rs

```diff
@@ -500,82 +499,51 @@ impl<C: sov_modules_api::Context> Evm<C> {
-                }
-            },
+        let block_number = match block_id {
+            Some(BlockId::Number(block_num)) => block_num,
             Some(BlockId::Hash(block_hash)) => {
                 let block_number = self
                     .get_block_number_by_block_hash(block_hash.block_hash, working_set)
                     .ok_or_else(|| EthApiError::UnknownBlockOrTxIndex)?;
+                BlockNumberOrTag::Number(block_number)
+            }
+            None => BlockNumberOrTag::Latest,
+        };
 
-                let block_env = BlockEnv::from(
-                    &self
-                        .get_sealed_block_by_number(
-                            Some(BlockNumberOrTag::Number(block_number)),
-                            working_set,
-                        )
-                        .unwrap()
-                        .expect("Block must be set"),
-                );
+        let (block_env, mut cfg_env) = {
+            let block = self
+                .get_sealed_block_by_number(Some(block_number), working_set)?
+                .ok_or(EthApiError::UnknownBlockNumber)?;
+            let block_env = BlockEnv::from(&block);
 
-                set_state_to_end_of_evm_block(block_number, working_set);
+            let cfg = self
+                .cfg
+                .get(working_set)
+                .expect("EVM chain config should be set");
+            let cfg_env = get_cfg_env(&block_env, cfg);
 
-                block_env
-            }
-            None => BlockEnv::from(
-                &self
-                    .get_sealed_block_by_number(Some(BlockNumberOrTag::Latest), working_set)
-                    .unwrap()
-                    .expect("Genesis block must be set"),
-            ),
+            (block_env, cfg_env)
         };
 
-        let cfg = self
-            .cfg
-            .get(working_set)
-            .expect("EVM chain config should be set");
-        let mut cfg_env = get_cfg_env(&block_env, cfg);
-
-        // set endpoint specific params
-        cfg_env.disable_eip3607 = true;
-        cfg_env.disable_base_fee = true;
-        // set higher block gas limit than usual
-        // but still cap it to prevent DoS
-        block_env.gas_limit = 100_000_000;
+        match block_number {
+            BlockNumberOrTag::Pending | BlockNumberOrTag::Latest => {}
+            _ => {
+                set_state_to_end_of_evm_block(block_env.number, working_set);
+            }
+        };
 
         let mut evm_db = self.get_db(working_set);
-        let mut tx_env = prepare_call_env(
+        let tx_env = prepare_call_env(
             &block_env,
-            request.clone(),
-            Some(
-                evm_db
-                    .basic(request.from.unwrap_or_default())
-                    .unwrap()
-                    .unwrap_or_default()
-                    .balance,
-            ),
+            &mut cfg_env,
+            &mut request.clone(),
```

Comment:
there might be a logical mistake here:

you are passing request as mutable reference, and in prepare_call_env, it is indeed mutated. 

however, you're passing the clone of request, which you keep using the original


my suggestion would be:


```rust
let cap_balance = evm_db.basic(request.from.unwrap_or_default()).....;

let tx_env = prepare_call_env(....., &mut request, cap_balance)?;
```

Reply:
It is indeed funky, fixed it by making prepare_call_env consume transaction request